### PR TITLE
planner: unify the behavior of prepare/execute limit to mysql

### PR DESCRIPTION
**Pull request by fzzf678**
### What problem does this PR solve?

TiDB's behavior in limit statement in Prepare/Execute is different from MySQL's.

MySQL
截屏2023-01-05 下午3 10 09

TiDB
截屏2023-01-05 下午3 15 03
even this, 
截屏2023-01-05 下午3 15 43
we'd better unify this behavior with MySQL.

Issue Number: ref #40219 

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -2041,17 +2041,80 @@ func getUintFromNode(ctx sessionctx.Context, n ast.Node) (uVal uint64, isNull bo
...
+		}
+		str, isNull, err := expression.GetStringFromConstant(ctx, param)
+		if err != nil {
+			return 0, false, false
+		}
+		if isNull {
+			return 0, true, true
+		}
+		val = str
+	default:
```
Comment (qw4990) [inline planner/core/logical_plan_builder.go:2071]: What if `limit 1+?`?

File: planner/core/logical_plan_builder.go [hunk 2]
```diff
@@ -2041,17 +2041,80 @@ func getUintFromNode(ctx sessionctx.Context, n ast.Node) (uVal uint64, isNull bo
...
+			return uint64(v), false, true
+		}
+	case string:
+		sc := ctx.GetSessionVars().StmtCtx
+		uVal, err := types.StrToUint(sc, v, false)
+		if err != nil {
+			return 0, false, false
+		}
+		return uVal, false, true
+	}
```
Comment (qw4990) [inline planner/core/logical_plan_builder.go:2088]: Should we add `case int` here?

File: planner/core/logical_plan_builder.go [hunk 3]
```diff
@@ -2041,17 +2041,80 @@ func getUintFromNode(ctx sessionctx.Context, n ast.Node) (uVal uint64, isNull bo
...
+	return 0, false, false
+}
+
+// check param type for plan cache, only allow int now
+// eg: set @a = 1;
+func checkParamTypeExpected(param *driver.ParamMarkerExpr) bool {
+	val := param.GetValue()
+	switch v := val.(type) {
+	case int64:
```
Comment (qw4990) [inline planner/core/logical_plan_builder.go:2097]: Should we add `case int` here as well?

File: planner/core/logical_plan_builder.go [hunk 4]
```diff
@@ -2041,17 +2041,80 @@ func getUintFromNode(ctx sessionctx.Context, n ast.Node) (uVal uint64, isNull bo
 	return 0, false, false
 }
 
+// getUintFromLimitNode gets uint64 value from ast.Node.
+// For ordinary statement, node should be uint64 constant value.
+// For prepared statement, node is string. We should convert it to uint64.
+func getUintFromLimitNode(ctx sessionctx.Context, n ast.Node) (uVal uint64, isNull bool, isExpectedType bool) {
```
Comment (qw4990) [inline planner/core/logical_plan_builder.go:2047]: Can we merge this function with `getUintFromNode`? They are too similar. We can add a new argument on `getUintFromNode` to indicate whether the returned value must be UINT like `getUintFromNode(ctx, n, mustUint)`.

Comment (fzzf678) [reply planner/core/logical_plan_builder.go:2047]: yeah, we can do this

File: planner/core/logical_plan_builder.go (hunk 2 above)
Comment (fzzf678) [reply planner/core/logical_plan_builder.go:2088]: I think the `Datum` 's value are set and get by unify functions, but there seems no `SetInt()` or `GetInt()` function exist, and the `Datum` struct has no `int` type members except `interface{}` can be used to save。So i guess there is no need to `case int `

Comment (fzzf678) [reply planner/core/logical_plan_builder.go]: And this function is clone from `getUintFromNode`, this function has also no `int` case

File: planner/core/logical_plan_builder.go [hunk 5]
```diff
@@ -2047,17 +2052,32 @@ func getUintFromNode(ctx sessionctx.Context, n ast.Node) (uVal uint64, isNull bo
...
 	if limit.Count != nil {
-		count, _, isExpectedType = getUintFromNode(ctx, limit.Count)
+		count, _, isExpectedType = getUintFromNode(ctx, limit.Count, true)
 		if !isExpectedType {
 			return 0, 0, ErrWrongArguments.GenWithStackByArgs("LIMIT")
 		}
 	}
 	if limit.Offset != nil {
-		offset, _, isExpectedType = getUintFromNode(ctx, limit.Offset)
+		offset, _, isExpectedType = getUintFromNode(ctx, limit.Offset, true)
```
Comment (Reminiscent) [inline planner/core/logical_plan_builder.go:2080]: Please add some test cases for this situation.

**Review by Reminiscent — APPROVED** [review]

**Review by qw4990 — APPROVED** [review]

**Pull request merged**